### PR TITLE
Fix the sigmoid cross entropy with logits.

### PR DESCRIPTION
**Pull request by xiejw**
Fix the sigmoid cross entropy with logits. 

Particularly,
1. logits should be wrapped into a sigmoid. The original impl actually assumes the logits are outputs of sigmoid.
2. The mean loss should be reduced by number of elements not batch size. 
3. Use a numerical stable implementation to avoid Nan.

### CODE REVIEW

File: Sources/DeepLearning/Loss.swift [hunk 1]
```diff
@@ -74,13 +74,17 @@ func _vjpSoftmaxCrossEntropy<Scalar: TensorFlowFloatingPoint>(
 /// Computes the sigmoid cross entropy (binary cross entropy) between logits and labels.
 ///
 /// - Parameters:
-///   - logits: Single continuous values from `0` to `1`.
+///   - logits: Unscaled value from a neural network.
 ///   - labels: Integer values that correspond to the correct output.
-@differentiable
+@differentiable(wrt: logits)
 public func sigmoidCrossEntropy<Scalar: TensorFlowFloatingPoint>(
     logits: Tensor<Scalar>, labels: Tensor<Scalar>
```
Comment (saeta) [inline Sources/DeepLearning/Loss.swift]: Should there be 2 generic types, so that `labels` can be a Tensor of a non-`TensorFlowFloatingPoint` type?

File: Sources/DeepLearning/Loss.swift [hunk 2]
```diff
@@ -74,13 +74,17 @@ func _vjpSoftmaxCrossEntropy<Scalar: TensorFlowFloatingPoint>(
...
 ) -> Tensor<Scalar> {
-    let loss = labels * log(logits) +
-        (Tensor<Scalar>(1) - labels) * log(Tensor<Scalar>(1) - logits)
-    return -loss.mean(alongAxes: 0).sum()
+    // This numerical stable implementation is based on tf.nn.sigmoid_cross_entropy_with_logits.
+
+    let maxLogitsWithZero = max(logits, Tensor(0))
+    // TODO(SR-10061): Use abs directly once abc is differentiable.
```
Comment (saeta) [inline Sources/DeepLearning/Loss.swift]: `s/SR-10061/TF-349/` (I moved the bug you filed to the right project: https://bugs.swift.org/browse/TF-349)

Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]:
```suggestion
    // TODO(SR-10061): Use abs directly once `abs` is differentiable.
```

File: Sources/DeepLearning/Loss.swift [hunk 3]
```diff
@@ -74,13 +74,17 @@ func _vjpSoftmaxCrossEntropy<Scalar: TensorFlowFloatingPoint>(
 /// Computes the sigmoid cross entropy (binary cross entropy) between logits and labels.
 ///
 /// - Parameters:
-///   - logits: Single continuous values from `0` to `1`.
+///   - logits: Unscaled value from a neural network.
```
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]:
```suggestion
///   - logits: The unscaled output of a neural network.
```

File: Sources/DeepLearning/Loss.swift (hunk 1 above)
Comment (rxwei) [reply Sources/DeepLearning/Loss.swift]: I don't think so. Labels need to be multiplied with logits (see line 88).

Comment (rxwei) [reply Sources/DeepLearning/Loss.swift]: I don't think so. Labels need to be multiplied with logits (see line 88).

**Comment by lattner** [comment]
Very nice Jianwei!

File: Sources/DeepLearning/Loss.swift (hunk 2 above)
Comment (xiejw) [reply Sources/DeepLearning/Loss.swift]: Done.

File: Sources/DeepLearning/Loss.swift [hunk 4]
```diff
@@ -74,13 +74,17 @@ func _vjpSoftmaxCrossEntropy<Scalar: TensorFlowFloatingPoint>(
...
 ) -> Tensor<Scalar> {
-    let loss = labels * log(logits) +
-        (Tensor<Scalar>(1) - labels) * log(Tensor<Scalar>(1) - logits)
-    return -loss.mean(alongAxes: 0).sum()
+    // This numerical stable implementation is based on tf.nn.sigmoid_cross_entropy_with_logits.
+
+    let maxLogitsWithZero = max(logits, Tensor(0))
+    // TODO(SR-10061): Use abs directly once `abs` is differentiable.
+    let absOfLogitsbsl = max(logits, -logits)
+    let loss =  maxLogitsWithZero - logits * labels + log(1 + exp(-absOfLogitsbsl))
```
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]:
```suggestion
    let loss = maxLogitsWithZero - logits * labels + log(1 + exp(-absOfLogitsbsl))
```
